test/fixtures/rules/std_rule_cases/L004.yml

```diff
@@ -80,6 +80,8 @@ indented_comments:
     FROM spam
 
 indented_comments_default_config:
+  # TODO partial fix - this requires some more attention
+  #  Not visible to bare eye but here tab indent is replaced with spaces
```

Comment:
I tried to fix this manually from the command line with the following command (which generates more verbose linter output):
```
sqlfluff fix test.sql --fixed-suffix FIXED --force --rules L004 -vvvv
```
The following messages were printed, so this is a legitimate case where the rule **finds** an issue but cannot **fix** the issue. We can remove the `fix_str` from this test case.
```
DEBUG      [L004] !! Violation Found: 'Incorrect indentation type found in file. The indent occurs after other text, so a manual fix is needed.'                                                        
DEBUG      [L004] !! Violation Found: 'Incorrect indentation type found in file. The indent occurs after other text, so a manual fix is needed.'                                                        
```

Reply:
Is it ok that fix is partial?

If I remove the fix_str here then this new assertion will fail:
```python
            # check that tests without a fail_str do no apply ANY fixes
             assert res == test_case.fail_str, "fix should not have modified the query"
```

Also, if we don't assert the result with fix_str any more, test coverage goes down.

I proposed instead adding a new test param to cover these cases, for example:
```yaml
  fail_after_fix: True
```
That way we can assert that the modifications after fix are as expected AND that it's still able to catch the remaining unfixable violation.

Reply:
I think it's OK that the fix is partial. It's not ideal but at least the test confirms that's what happens and it's what;'s expected.

I presume it consistently "fixes" to the same SQL after `fix` is run from the original SQL.

Reply:
> I presume it consistently "fixes" to the same SQL after fix is run from the original SQL?

Yes. That is asserted by the existing code. What the existing code doesn't check is that the check still fails on the fix_str. I propose adding a new param to state that + code to assert it.

Reply:
My other PR (#2657) would not help in this case. The rule is not buggy -- it's behaving as expected. I reviewed a few of these TODOs today and found two cases:
* The rule finds an issue but does not generate a fix. This is fine!
* The rule finds an issue and generates a fix, but the core linter does not apply it. There _may_ be possible improvements to make here in the future, but it's working as designed for the moment (I don't want to try and change this now).

There are several potential paths forward:
* `fail_after_fix`: This could be a bool value, as proposed. Another option would be to list the specific violations. Should we consider that?
* Use `-- noqa` to suppress these messages. If we implement #2337, then this would effectively be the same as `fail_after_fix`.

Any thoughts on which of these is better? I think I prefer the second, because it leverages existing or future linter functionality, rather than making the SQLFluff **test framework** more complex.

Reply:
Couldn't agree more – using noqa is a clever idea. However, does sqlfluff complain about redundant noqa comments? If not, then we would be still missing the assertion for catching the lint violation.

Reply:
I still prefer the (third?) option of going what did we have and confirming that’s the fix output. Even if a subsequent run might change it again.

To me that’s what happens currently and saves having to define and set new flags or noqa any code.

So the only changes needed are:
1. Confirm if we have a `fail_str` we should not have a `pass_str` nor a second `fail_str`. This is an easy mistake to make so would be good to capture this. A `fail_str` should either be on its own or have a single `fix_str`.
2. Similarly a `pass_str` should always be on its own.
3. If there is a `fail_str` but no `fix_str` then we should confirm the rule does not fix anything. If it does then it should be in the test in a `fix_str` IMHO and the fact it isn’t makes it an invalid test. As I say above this doesn’t have to be the “final fix” as would happen with multiple independent runs (though ideally it would be!) but should be deterministic of one run of the rule by itself.

I think those three things are valuable to have and would make the test suite more robust and less prone to errors that we see now. There’s value in that even if it’s not the full, deterministic, solution we’d all ideally like.

As I say, at some point in the future, we could take the next step and confirm the fix has settled and will not change on subsequent runs (I.e. is completely deterministic). But there’s a bit of work to be done first there before we can enforce that IMHO and not sure it’s worth adding flags to note these cases for now, while we work to get there, as would make the tests a bit more complex to understand and write for those that don’t know the full history here or what exactly those flags mean and why they are set on some tests and not others.

Reply:
I'm happy with your suggestions, @tunetheweb: it's often good to split work into smaller pieces so we can make steady progress and avoid getting stuck for a long time on one issue. We can go ahead and create a new issue for the remaining work so we don't forget it.

@juhoautio: About your question:
>does sqlfluff complain about redundant noqa comments? If not, then we would be still missing the assertion for catching the lint violation.

It doesn't catch this now, but see issue #2337 mentioned earlier: it's a proposed enhancement for that exact change. Someone could start work on that whenever they have time. Are you interested in working on that? Once we implement that, I think we'd be ready to work on the **new** issue we were discussing, i.e. enforcing that the `fix_str` passes the linter.

Reply:
I have some second thoughts about the generality of noqa for this purpose. Is it too theoretic or not, but if we want to ignore a remaining error on some row, we need to add the noqa comment already in the fail_str. If the remaining error is expected to happen on that same line, we won't be able to test that partial fix, because also the initial fix will be skipped due to the noqa marker on the line.

Reply:
>If the remaining error is expected to happen on that same line, we won't be able to test that partial fix

It's possible, but I think it will be unusual. If it does happen, in many cases, I think we could update the test case (possibly creating new ones) to test those errors separately.

I agree noqa probably isn't perfect, but it seems pretty close, so I'm in favor of trying it. If it doesn't work well, we can easily try something new, because the SQLFluff tests only affect developers, not users.